Refactor code for user-group based stream settings.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
This PR is for follow-ups mentioned in #23992.

- First commit is to add `require_system_group` parameter to `get_realm_user_groups_for_dropdown_list_widget` in `user_groups.ts`.
- Second commit is to refactor the backend code for `can_remove_subscribers_group` setting such that we can avoid duplication while adding more similar settings.
- Third commit adds a migration to fix `RealmAuditLog` objects.

<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on zerver/views/streams.py:
```diff
@@ -269,7 +269,9 @@ def update_stream_backend(
     ),
-    can_remove_subscribers_group_id: Optional[int] = REQ(json_validator=check_int, default=None),
+    can_remove_subscribers_group: Optional[int] = REQ(
+        "can_remove_subscribers_group_id", json_validator=check_int, default=None
```
I think for type reasons I'd prefer to keep the variable called `_id` if it's an integer, and compute a `setting_group_id_name` variable in the loop below.

timabbott commented on zerver/views/streams.py:
```diff
@@ -384,19 +386,27 @@ def update_stream_backend(
+                # Admins cannot change this setting for unsubscribed private streams.
+                raise JsonableError(_("Invalid stream ID"))
+
+            group_permission_setting_info = Stream.stream_permission_group_settings[setting_name]
```
Would it make more sense to just loop over `Stream.stream_permission_group_settings.items()` once, and just have a `if not hasattr(locals(), setting_group_id_name): continue` check at the top of the loop? 

I feel like having to reference this again in order to get the metadata feels a bit harder to read than `for setting_name, permissions_configuration in Stream.stream_permission_group_settings:` as the outer loop here.

timabbott replied:
The updated loop looks good; added this hunk to avoid the extra fetch:

```
diff --git a/zerver/views/streams.py b/zerver/views/streams.py
index d94914dbcb..7b98b0ec2c 100644
--- a/zerver/views/streams.py
+++ b/zerver/views/streams.py
@@ -399,14 +399,13 @@ def update_stream_backend(
                 raise JsonableError(_("Invalid stream ID"))
 
             user_group_id = request_settings_dict[setting_group_id_name]
-            group_permission_setting_info = Stream.stream_permission_group_settings[setting_name]
             user_group = access_user_group_for_setting(
                 user_group_id,
                 user_profile,
                 setting_name=setting_name,
-                require_system_group=group_permission_setting_info.require_system_group,
-                allow_internet_group=group_permission_setting_info.allow_internet_group,
-                allow_owners_group=group_permission_setting_info.allow_owners_group,
+                require_system_group=permissions_configuration.require_system_group,
+                allow_internet_group=permissions_configuration.allow_internet_group,
+                allow_owners_group=permissions_configuration.allow_owners_group,
             )
             do_change_stream_group_based_setting(
                 stream, setting_name, user_group, acting_user=user_profile
```

timabbott commented on zerver/migrations/0429_fix_audit_log_objects_for_group_based_stream_settings.py:
```diff
@@ -0,0 +1,46 @@
+
+def fix_audit_log_objects_for_group_based_stream_settings(
+    apps: StateApps, schema_editor: BaseDatabaseSchemaEditor
+) -> None:
```
Let's add a docstring here containing the content in the commit message -- that this adds the `property_name` field to any `STREAM_GROUP_BASED_SETTING_CHANGED` audit log entries that were created before the previous commit.

timabbott commented:
Overall this looks great; posted a couple comments with suggested tweaks that might make it feel cleaner and more readable.

sahil839 commented on zerver/views/streams.py:
```diff
@@ -384,19 +384,33 @@ def update_stream_backend(
+        request_settings_dict = locals()
+        setting_group_id_name = setting_name + "_id"
+
+        if setting_group_id_name not in request_settings_dict:  # nocoverage
```
Not sure if there is a nice way to get coverage for this.

timabbott replied:
Yeah, I don't see an obvious way to do it; seems OK for now.

sahil839 commented:
@timabbott updated the PR.
- Added docstring for the migration function.
- Fixed the function parameter to have `_id`.
- Updated the code to have `for setting_name, permissions_configuration in Stream.stream_permission_group_settings:` as outer loop.

timabbott commented on zerver/views/streams.py:
```diff
@@ -384,19 +384,33 @@ def update_stream_backend(
+        ] != getattr(stream, setting_name):
+            if sub is None and stream.invite_only:
+                # Admins cannot change this setting for unsubscribed private streams.
+                raise JsonableError(_("Invalid stream ID"))
```
We probably want to make this behavior controlled by a `GroupPermissionSetting` flag for whether this setting can be changed by unsubscribed administrators -- the idea being that the code here doesn't have special cases for particular settings. But that can be a follow-up.

Maybe `require_history_access_to_edit` would be a reasonable name for the flag?

timabbott commented:
Looks good, tagged to merge once CI finishes. Thanks @sahil839!
